[Feat-T3-65] BaseViewController 및 ViewModel protocol 구현

choijungp opened the pull request:
## 🌁 Background
<!-- 해당 PR을 작성하게 된 이유를 적어주세요. -->

ViewModel과 ViewController를 정형화 하기 위해 공통 `ViewModel` 프로토콜과 `BaseViewController`를 정의했습니다.

이를 통해 각 ViewController가 일관된 방식으로 ViewModel을 주입받고, 속성 설정(configureAttribute), 레이아웃 구성(configureLayout), 바인딩(bind)을 수행할 수 있도록 구조를 생성하고자 했습니다.

또한 네트워크 통신 테스트를 위해 HealthCheck 데이터를 ViewModel → ViewController로 전달하고 화면에 표시하는 로직을 검증하는 과정도 진행하였습니다.


## 📱 Screenshot 

<img src="https://github.com/user-attachments/assets/b7d5d115-08f8-43db-93ff-9b1fb0d62e38" width="250"/>


## 👩‍💻 Contents
<!-- 작업 내용을 적어주세요 -->
- ViewModel protocol 구현
- BaseViewController 구현 (모든 ViewController에서 BaseViewController를 상속받아 구현해주세요.)
- Test를 위한 객체들 수정 (HomeViewController, TestUseCase, TestRepository 등 ···)


## ✅ Testing
<!-- 테스트 방법을 적어주세요 -->
App 빌드 후 HomeViewController에 HealthCheck로 전달받은 값의 message를 띄워줬습니다.


## 📝 Review Note
<!-- PR과정에서 든 생각이나 개선할 내용이 있다면 적어주세요. -->
- NetworkService, Repository 단에서는 `async/await` 기반으로 네트워크 요청을 처리하고,
- UseCase → ViewModel → ViewController로 데이터가 전달되는 흐름은 `Combine`을 사용해 바인딩했습니다.

- ViewModel은 `Input-Output-Action` 구조로 설계하였습니다.  
  - **Input**: 이벤트(유저 액션 등)를 정의
  - **Output**: 화면에 필요한 데이터를 `Publisher` 형태로 노출하며  
  - **Action**: Input을 받아 내부 로직을 수행하는 방식으로 구성했습니다.


## 📣 Related Issue
<!-- 관련 이슈를 적어주세요. -->
- close #T3-65

choijungp committed:
Feat: ViewModel protocol 구현 (#T3-65)

choijungp committed:
Feat: ViewModel protocol 구현 (#T3-65)

choijungp committed:
Feat: BaseViewController 구현 (#T3-56)

choijungp committed:
Feat: BaseViewController 구현 (#T3-56)

choijungp committed:
Refactor: Test를 위한 HomeViewController 및 Test 객체들 수정

choijungp committed:
Refactor: Test를 위한 HomeViewController 및 Test 객체들 수정

generalban commented on Projects/Presentation/Sources/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,67 @@
+
+    private let label = UILabel()
+
+    public override init(viewModel: HomeViewModel) {
```
상위 클래스인 BaseViewController에서 이미 viewModel을 초기화하고 있지 않나요?

choijungp replied:
넵! `BaseViewController` viewModel public으로 수정 후 `HomeViewController` viewModel 프로퍼티 삭제하겠습니다.

generalban commented on Projects/Presentation/Sources/Home/HomeViewController.swift:
```diff
@@ -0,0 +1,67 @@
+import Then
+import Shared
+
+public final class HomeViewController: BaseViewController<HomeViewModel> {
```
BaseViewController가 이미 제네릭 타입 T로 viewModel을 선언하고 있어요 제가 생각한게 맞다면,,,

choijungp committed:
Feat: Design System 구현 (#T3-64)

* Feat: Gray 색상 Asset 추가 (#T3-64)

* Feat: BitnagilColor Enum 정의 (#T3-64)

* Feat: Pretendard 폰트 파일 추가

* Feat: BitnagilFont Enum 정의 (#T3-64)

- BitnagilFont Enum 정의
- FontName Enum 정의
- UIFont extension 구현

* Chore: Project 파일 수정

- 각 모듈의 infoPlist 설정 제거
- App 단의 info.plist 파일 사용하도록 수정

choijungp committed:
Refactor: BaseViewController의 viewModel 접근 제한자 수정 및 중복 제거

choijungp committed:
Refactor: BaseViewController의 viewModel 접근 제한자 수정 및 중복 제거

choijungp committed:
Chore: develop 변경사항 반영

choijungp committed:
Refactor: HomeViewController에 디자인 시스템 적용

generalban reviewed (approved):
구조 잘 정리되어 있고, DI 흐름부터 리소스 추가까지 깔끔하게 정리된 것 같습니다
고생 많으셨습니다! 굿입니다!
